server/catgenome/src/main/java/com/epam/catgenome/manager/reference/MotifSearchManager.java

```diff
@@ -143,27 +145,69 @@ private MotifSearchResult searchChromosomeMotifs(final MotifSearchRequest reques
                 .build();
     }
 
-    private MotifSearchResult searchWholeGenomeMotifs(final MotifSearchRequest motifSearchRequest) {
-        final Chromosome chr = loadChrById(motifSearchRequest.getReferenceId(), motifSearchRequest.getChromosomeId());
-        int start = motifSearchRequest.getStartPosition() == null
-                ? 0
-                : motifSearchRequest.getStartPosition();
-        int end = motifSearchRequest.getEndPosition() == null
-                ? chr.getSize()
-                : motifSearchRequest.getEndPosition();
-
+    private MotifSearchResult searchWholeGenomeMotifs(final MotifSearchRequest request) {
+        final int startPageSize = request.getPageSize() == null ? DEFAULT_PAGE_SIZE : request.getPageSize();
+        int start = request.getStartPosition() == null ? 0 : request.getStartPosition();
+        Chromosome chromosome = loadChrById(request.getReferenceId(), request.getChromosomeId());
+        long chrId = chromosome.getId();
+        int end = request.getEndPosition() == null ? chromosome.getSize() : request.getEndPosition();
+        if (end < chromosome.getSize()) {
+            return searchRegionMotifs(
+                    MotifSearchRequest.builder()
+                            .referenceId(request.getReferenceId())
+                            .strand(request.getStrand())
+                            .pageSize(startPageSize)
+                            .searchType(MotifSearchType.REGION)
+                            .motif(request.getMotif())
+                            .chromosomeId(chrId)
+                            .startPosition(start)
+                            .endPosition(end)
+                            .build());
+        }
+        final List<Chromosome> chromosomes = getChromosomesOfGenome(request.getReferenceId());
+        final List<Motif> motifs = new ArrayList<>();
+        int pageSize = startPageSize;
+        while (chromosome != null) {
+            chrId = chromosome.getId();
+            end = chromosome.getSize();
+            List<Motif> result = searchChromosomeMotifs(
+                    MotifSearchRequest.builder()
+                            .referenceId(request.getReferenceId())
+                            .strand(request.getStrand())
+                            .pageSize(pageSize)
+                            .searchType(MotifSearchType.CHROMOSOME)
+                            .motif(request.getMotif())
+                            .chromosomeId(chrId)
+                            .startPosition(start)
+                            .endPosition(end)
+                            .build())
+                    .getResult();
+            motifs.addAll(result.stream().limit(pageSize).collect(Collectors.toList()));
+            pageSize -= result.size();
```

Comment:
instead of introducing `startPageSize` value we can always calculate `pageSize - motif.size()` it will be more readable